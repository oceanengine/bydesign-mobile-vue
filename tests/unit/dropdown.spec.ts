import { mount, later } from '../index';

function renderWrapper(options = {} as any) {
  return mount({
    template: `
      <byted-dropdown
        :direction="direction"
        :close-on-click-outside="closeOnClickOutside">
        <byted-dropdown-item v-model="value" :title="title" :options="options" />
        <byted-dropdown-item v-model="value" :title="title" :options="options" />
      </byted-dropdown>
    `,
    data() {
      return {
        value: options.value || 0,
        title: options.title || '',
        direction: options.direction || 'down',
        closeOnClickOutside: options.closeOnClickOutside,
        options: [
          { text: 'A', value: 0 },
          { text: 'B', value: 1 },
        ],
      };
    },
  });
}

describe('dropdown', () => {
  test('show dropdown item', async () => {
    const wrapper = renderWrapper();

    await later();

    const titles = wrapper.findAll('.byted-dropdown__item');

    await titles.at(0).trigger('click');
    expect(wrapper).toMatchSnapshot();

    await titles.at(1).trigger('click');
    expect(wrapper).toMatchSnapshot();

    await titles.at(1).trigger('click');
    expect(wrapper).toMatchSnapshot();
  });
  test('close-on-click-outside', async () => {
    const wrapper = renderWrapper({
      closeOnClickOutside: true,
    });

    await later();

    const titles = wrapper.findAll('.byted-dropdown__item');

    await titles.at(0).trigger('click');
    await document.body.click();
    expect(wrapper).toMatchSnapshot();
  });

  test('disable close-on-click-outside', async () => {
    const wrapper = renderWrapper({
      closeOnClickOutside: false,
    });

    await later();

    const titles = wrapper.findAll('.byted-dropdown__item');

    await titles.at(0).trigger('click');
    await document.body.click();
    expect(wrapper).toMatchSnapshot();
  });

  test('direction up', async () => {
    const wrapper = renderWrapper({
      direction: 'up',
    });

    await later();
    expect(wrapper).toMatchSnapshot();

    const titles = wrapper.findAll('.byted-dropdown__item');
    await titles.at(0).trigger('click');
    expect(wrapper).toMatchSnapshot();
  });

  test('click option', async () => {
    const wrapper = renderWrapper();

    await later();

    const titles = wrapper.findAll('.byted-dropdown__item');
    await titles.at(0).trigger('click');

    const options = wrapper.findAll('.byted-dropdown-item .byted-cell');
    await options.at(1).trigger('click');

    expect(wrapper).toMatchSnapshot();
  });
  test('title prop', async () => {
    const wrapper = renderWrapper({ title: 'Title' });
    await later();
    expect(wrapper).toMatchSnapshot();
  });
  test('didn`t find matched option', async () => {
    const wrapper = renderWrapper({ value: -1 });
    await later();
    expect(wrapper).toMatchSnapshot();
  });

  test('destroy one item', async () => {
    const wrapper = mount({
      template: `
      <byted-dropdown>
        <byted-dropdown-item v-if="render" v-model="value" :options="options" />
        <byted-dropdown-item v-model="value" :options="options" />
      </byted-dropdown>
    `,
      data() {
        return {
          value: 0,
          render: true,
          options: [
            { text: 'A', value: 0 },
            { text: 'B', value: 1 },
          ],
        };
      },
    });

    await later();
    wrapper.setData({ render: false });
    expect(wrapper).toMatchSnapshot();
  });

  test('disable dropdown item', async () => {
    const wrapper = mount({
      template: `
      <byted-dropdown>
        <byted-dropdown-item disabled v-model="value" :options="options" />
      </byted-dropdown>
    `,
      data() {
        return {
          value: 0,
          options: [
            { text: 'A', value: 0 },
            { text: 'B', value: 1 },
          ],
        };
      },
    });
    await later();
    const titles = wrapper.findAll('.byted-dropdown__item');
    await titles.at(0).trigger('click');
    expect(wrapper).toMatchSnapshot();
  });
  test('change event', async () => {
    const onChange = jest.fn();

    const wrapper = mount({
      template: `
      <byted-dropdown>
        <byted-dropdown-item  v-model="value" :options="options"  @change="onChange" />
        <byted-dropdown-item  v-model="value" :options="options" />
      </byted-dropdown>
    `,
      data() {
        return {
          value: 0,
          options: [
            { text: 'A', value: 0 },
            { text: 'B', value: 1 },
          ],
        };
      },
      methods: {
        onChange,
      },
    });

    await later();

    const titles = wrapper.findAll('.byted-dropdown__item');
    await titles.at(0).trigger('click');
    await later();
    const options = wrapper.findAll('.byted-dropdown-item .byted-cell');
    await options.at(0).trigger('click');

    expect(onChange).toHaveBeenCalledTimes(0);

    await options.at(1).trigger('click');
    expect(onChange).toHaveBeenCalledWith(1);
    expect(onChange).toHaveBeenCalledTimes(1);
  });
  test('title slot', () => {
    const wrapper = mount({
      template: `
      <byted-dropdown>
        <van-dropdown-item>
          <template #title>
            Custom Title
          </template>
        </van-dropdown-item>
      </byted-dropdown>
    `,
    });

    expect(wrapper).toMatchSnapshot();
  });
});
