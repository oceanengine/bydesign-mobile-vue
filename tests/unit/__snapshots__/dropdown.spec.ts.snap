// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropdown click option 1`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">B</div></span></div>
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">B</div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px);">
    <transition-stub name="byted-popup-slide-top" class="byted-dropdown-item__content" style="z-index: 2011;">
      <div class="byted-popup byted-popup--top" style="transition-duration: 0.2s; display: none;">
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(51, 51, 51);">
          <div class="byted-cell__title"><span>A</span></div>
        </div>
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(2, 120, 255);">
          <div class="byted-cell__title"><span>B</span></div>
          <div class="byted-cell__value"><i class="byted-icon byted-icon--check byted-dropdown-item__icon" style="fill: #0278FF; width: 0.8rem; height: 0.8rem;">test-file-stub</i></div>
        </div>
      </div>
    </transition-stub>
  </div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px); display: none;">
    <!---->
  </div>
</div>
`;

exports[`dropdown close-on-click-outside 1`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item"><span class="byted-dropdown__title byted-dropdown__title--down" style="color: rgb(2, 120, 255);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px);">
    <div class="byted-overlay byted-fade-enter byted-fade-enter-active" style="z-index: 2004; position: absolute; transition-duration: 0.2s;"></div>
    <transition-stub name="byted-popup-slide-top" class="byted-dropdown-item__content" style="z-index: 2005;">
      <div class="byted-popup byted-popup--top" style="transition-duration: 0.2s;">
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(2, 120, 255);">
          <div class="byted-cell__title"><span>A</span></div>
          <div class="byted-cell__value"><i class="byted-icon byted-icon--check byted-dropdown-item__icon" style="fill: #0278FF; width: 0.8rem; height: 0.8rem;">test-file-stub</i></div>
        </div>
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(51, 51, 51);">
          <div class="byted-cell__title"><span>B</span></div>
        </div>
      </div>
    </transition-stub>
  </div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px); display: none;">
    <!---->
  </div>
</div>
`;

exports[`dropdown destroy one item 1`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px); display: none;">
    <!---->
  </div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px); display: none;">
    <!---->
  </div>
</div>
`;

exports[`dropdown didn\`t find matched option 1`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis"></div></span></div>
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis"></div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px); display: none;">
    <!---->
  </div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px); display: none;">
    <!---->
  </div>
</div>
`;

exports[`dropdown direction up 1`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item"><span class="byted-dropdown__title byted-dropdown__title--down" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown__item"><span class="byted-dropdown__title byted-dropdown__title--down" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--up" style="z-index: 10; height: calc(100vh - 0px); transform: translate3d(0, -100%, 0); display: none;">
    <!---->
  </div>
  <div class="byted-dropdown-item byted-dropdown-item--up" style="z-index: 10; height: calc(100vh - 0px); transform: translate3d(0, -100%, 0); display: none;">
    <!---->
  </div>
</div>
`;

exports[`dropdown direction up 2`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(2, 120, 255);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown__item"><span class="byted-dropdown__title byted-dropdown__title--down" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--up" style="z-index: 10; height: calc(100vh - 768px); transform: translate3d(0, -100%, 0);">
    <div class="byted-overlay byted-fade-enter-active byted-fade-enter-to" style="z-index: 2008; position: absolute; transition-duration: 0.2s;"></div>
    <transition-stub name="byted-popup-slide-bottom" class="byted-dropdown-item__content" style="z-index: 2009;">
      <div class="byted-popup byted-popup--bottom" style="transition-duration: 0.2s;">
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(2, 120, 255);">
          <div class="byted-cell__title"><span>A</span></div>
          <div class="byted-cell__value"><i class="byted-icon byted-icon--check byted-dropdown-item__icon" style="fill: #0278FF; width: 0.8rem; height: 0.8rem;">test-file-stub</i></div>
        </div>
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(51, 51, 51);">
          <div class="byted-cell__title"><span>B</span></div>
        </div>
      </div>
    </transition-stub>
  </div>
  <div class="byted-dropdown-item byted-dropdown-item--up" style="z-index: 10; height: calc(100vh - 768px); transform: translate3d(0, -100%, 0); display: none;">
    <!---->
  </div>
</div>
`;

exports[`dropdown disable close-on-click-outside 1`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item"><span class="byted-dropdown__title byted-dropdown__title--down" style="color: rgb(2, 120, 255);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px);">
    <div class="byted-overlay byted-fade-enter-active" style="z-index: 2006; position: absolute; transition-duration: 0.2s;"></div>
    <transition-stub name="byted-popup-slide-top" class="byted-dropdown-item__content" style="z-index: 2007;">
      <div class="byted-popup byted-popup--top" style="transition-duration: 0.2s;">
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(2, 120, 255);">
          <div class="byted-cell__title"><span>A</span></div>
          <div class="byted-cell__value"><i class="byted-icon byted-icon--check byted-dropdown-item__icon" style="fill: #0278FF; width: 0.8rem; height: 0.8rem;">test-file-stub</i></div>
        </div>
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(51, 51, 51);">
          <div class="byted-cell__title"><span>B</span></div>
        </div>
      </div>
    </transition-stub>
  </div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px); display: none;">
    <!---->
  </div>
</div>
`;

exports[`dropdown disable dropdown item 1`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item byted-dropdown__item--disabled"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px); display: none;">
    <!---->
  </div>
</div>
`;

exports[`dropdown show dropdown item 1`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item"><span class="byted-dropdown__title byted-dropdown__title--down" style="color: rgb(2, 120, 255);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px);">
    <div class="byted-overlay byted-fade-enter byted-fade-enter-active" style="z-index: 2000; position: absolute; transition-duration: 0.2s;"></div>
    <transition-stub name="byted-popup-slide-top" class="byted-dropdown-item__content" style="z-index: 2001;">
      <div class="byted-popup byted-popup--top" style="transition-duration: 0.2s;">
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(2, 120, 255);">
          <div class="byted-cell__title"><span>A</span></div>
          <div class="byted-cell__value"><i class="byted-icon byted-icon--check byted-dropdown-item__icon" style="fill: #0278FF; width: 0.8rem; height: 0.8rem;">test-file-stub</i></div>
        </div>
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(51, 51, 51);">
          <div class="byted-cell__title"><span>B</span></div>
        </div>
      </div>
    </transition-stub>
  </div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px); display: none;">
    <!---->
  </div>
</div>
`;

exports[`dropdown show dropdown item 2`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown__item"><span class="byted-dropdown__title byted-dropdown__title--down" style="color: rgb(2, 120, 255);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px);">
    <transition-stub name="byted-popup-slide-top" class="byted-dropdown-item__content" style="z-index: 2001;">
      <div class="byted-popup byted-popup--top" style="transition-duration: 0s; display: none;">
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(2, 120, 255);">
          <div class="byted-cell__title"><span>A</span></div>
          <div class="byted-cell__value"><i class="byted-icon byted-icon--check byted-dropdown-item__icon" style="fill: #0278FF; width: 0.8rem; height: 0.8rem;">test-file-stub</i></div>
        </div>
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(51, 51, 51);">
          <div class="byted-cell__title"><span>B</span></div>
        </div>
      </div>
    </transition-stub>
  </div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px);">
    <div class="byted-overlay byted-fade-enter byted-fade-enter-active" style="z-index: 2002; position: absolute; transition-duration: 0.2s;"></div>
    <transition-stub name="byted-popup-slide-top" class="byted-dropdown-item__content" style="z-index: 2003;">
      <div class="byted-popup byted-popup--top" style="transition-duration: 0.2s;">
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(2, 120, 255);">
          <div class="byted-cell__title"><span>A</span></div>
          <div class="byted-cell__value"><i class="byted-icon byted-icon--check byted-dropdown-item__icon" style="fill: #0278FF; width: 0.8rem; height: 0.8rem;">test-file-stub</i></div>
        </div>
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(51, 51, 51);">
          <div class="byted-cell__title"><span>B</span></div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`dropdown show dropdown item 3`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">A</div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px);">
    <transition-stub name="byted-popup-slide-top" class="byted-dropdown-item__content" style="z-index: 2001;">
      <div class="byted-popup byted-popup--top" style="transition-duration: 0s; display: none;">
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(2, 120, 255);">
          <div class="byted-cell__title"><span>A</span></div>
          <div class="byted-cell__value"><i class="byted-icon byted-icon--check byted-dropdown-item__icon" style="fill: #0278FF; width: 0.8rem; height: 0.8rem;">test-file-stub</i></div>
        </div>
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(51, 51, 51);">
          <div class="byted-cell__title"><span>B</span></div>
        </div>
      </div>
    </transition-stub>
  </div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px);">
    <div class="byted-overlay byted-fade-leave byted-fade-leave-active" style="z-index: 2002; position: absolute; transition-duration: 0.2s;"></div>
    <transition-stub name="byted-popup-slide-top" class="byted-dropdown-item__content" style="z-index: 2003;">
      <div class="byted-popup byted-popup--top" style="transition-duration: 0.2s; display: none;">
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(2, 120, 255);">
          <div class="byted-cell__title"><span>A</span></div>
          <div class="byted-cell__value"><i class="byted-icon byted-icon--check byted-dropdown-item__icon" style="fill: #0278FF; width: 0.8rem; height: 0.8rem;">test-file-stub</i></div>
        </div>
        <div class="byted-cell byted-cell--clickable byted-cell--large" style="color: rgb(51, 51, 51);">
          <div class="byted-cell__title"><span>B</span></div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`dropdown title prop 1`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">Title</div></span></div>
  <div class="byted-dropdown__item"><span class="byted-dropdown__title" style="color: rgb(51, 51, 51);"><div class="byted-ellipsis">Title</div></span></div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px); display: none;">
    <!---->
  </div>
  <div class="byted-dropdown-item byted-dropdown-item--down" style="z-index: 10; height: calc(100vh - 0px); display: none;">
    <!---->
  </div>
</div>
`;

exports[`dropdown title slot 1`] = `
<div class="byted-dropdown byted-hairline--top-bottom">
  <van-dropdown-item></van-dropdown-item>
</div>
`;
